--- serefpolicy-2.6.4/policy/modules/services/postgresql.te	2007-05-08 03:51:01.000000000 +0900
+++ serefpolicy-2.6.4.sepgsql/policy/modules/services/postgresql.te	2007-06-17 11:54:43.000000000 +0900
@@ -175,16 +175,6 @@ optional_policy(`
 ')
 
 ifdef(`TODO',`
-ifdef(`targeted_policy', `', `
-bool allow_user_postgresql_connect false;
-
-if (allow_user_postgresql_connect) {
-# allow any user domain to connect to the database server
-allow userdomain postgresql_t:unix_stream_socket connectto;
-allow userdomain postgresql_var_run_t:sock_file write;
-allow userdomain postgresql_tmp_t:sock_file write;
-}
-')
 ifdef(`distro_debian', `
 	init_exec_script_files(postgresql_t)
 	# gross hack
--- serefpolicy-2.6.4/policy/modules/system/userdomain.te	2007-06-17 11:49:53.000000000 +0900
+++ serefpolicy-2.6.4.sepgsql/policy/modules/system/userdomain.te	2007-06-17 11:54:43.000000000 +0900
@@ -31,6 +31,13 @@ gen_tunable(allow_user_mysql_connect,fal
 
 ## <desc>
 ## <p>
+## Allow users to connect to PostgreSQL
+## </p>
+## </desc>
+gen_tunable(allow_user_postgresql_connect,false)
+
+## <desc>
+## <p>
 ## Allow regular users direct mouse access
 ## </p>
 ## </desc>
--- serefpolicy-2.6.4/policy/modules/system/userdomain.if	2007-06-17 11:49:53.000000000 +0900
+++ serefpolicy-2.6.4.sepgsql/policy/modules/system/userdomain.if	2007-06-17 11:54:43.000000000 +0900
@@ -909,6 +909,12 @@ template(`userdom_common_user_template',
 	')
 
 	optional_policy(`
+		tunable_policy(`allow_user_postgresql_connect',`
+			postgresql_stream_connect($1_t)
+		')
+	')
+
+	optional_policy(`
 		nscd_socket_use($1_t)
 	')
 
